### PR TITLE
Add memorymap support to RP2 port

### LINKED ISSUE #8334: RP2040 lacks raw memory map access
The RP2040 port lacks support for raw memory map access. Implementing the core memorymap module will provide access to RP2040 registers that are not otherwise available in CircuitPython, such as setting pad hysteresis. Let me know if this is desirable and I'll code a pull.

### PR DESCRIPTION
**Pull request by eightycc**
Resolves #8334 by adding memorymap support to the RP2 port. Also makes these additional changes:

- Changed memorymap hal function prototypes to use size_t rather than uint32_t arguments, allowing prototypes to work with different port word sizes.
- Changes memorymap.AddressRange() start argument from signed to an unsigned int to allow ranges to be specified anywhere in the entire port address map. Added support to the runtime for unsigned int arguments.
- Added wrap-around checking for memorymap.AddressRange() arguments.
- Tweaked Python commentary embedded in memorymap to reflect reality of single operation accesses.

### CODE REVIEW

**Comment by eightycc** [comment]
CI found some breakage. I entirely missed the fact that there are multiple runtimes. Working...

**Comment by eightycc** [comment]
@tannewt over to you for review. Thanks!

**Comment by eightycc** [comment]
After tuning in to the meeting today my understanding of 8.2.x vs. main branch has improved. @tannewt would you like me to rework this pull request for 8.2.x?

**Comment by tannewt** [comment]
> After tuning in to the meeting today my understanding of 8.2.x vs. main branch has improved. @tannewt would you like me to rework this pull request for 8.2.x?

Only if you want to see it in an 8.x release. Otherwise it's fine on main.

File: ports/raspberrypi/common-hal/memorymap/AddressRange.c [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    {(uint8_t *)ROM_BASE,           0x00004000, ROM},        // boot ROM
+    {(uint8_t *)XIP_BASE,           0x00100000, XIP},        // XIP normal cache operation
+    {(uint8_t *)XIP_NOALLOC_BASE,   0x00100000, XIP},        // XIP check for hit, no update on miss
+    {(uint8_t *)XIP_NOCACHE_BASE,   0x00100000, XIP},        // XIP don't check for hit, no update on miss
+    {(uint8_t *)XIP_NOCACHE_NOALLOC_BASE, 0x00100000, XIP},  // XIP bypass cache completely
+    {(uint8_t *)XIP_CTRL_BASE,      0x00004000, IO},         // XIP control registers
+    {(uint8_t *)XIP_SRAM_BASE,      0x00004000, SRAM},       // XIP SRAM 16KB XIP cache
+    {(uint8_t *)XIP_SSI_BASE,       0x00004000, IO},         // XIP SSI registers
+    {(uint8_t *)SRAM_BASE,          0x00042000, SRAM},       // SRAM 256KB striped plus 16KB contiguous
+    {(uint8_t *)SRAM0_BASE,         0x00040000, SRAM},       // SRAM0 to SRAM3 256KB non-striped
```
Comment (tannewt) [inline ports/raspberrypi/common-hal/memorymap/AddressRange.c:47]: I'm a little wary of allowing SRAM access where CP is maintaining the heap. Do you *need* this? I'd prefer to be conservative to start and add areas as we have specific uses.

File: ports/raspberrypi/common-hal/memorymap/AddressRange.c [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    switch (self->type) {
+        case SRAM:
+            // Writes to SRAM may be arbitrary length and alignment. We use memcpy() which
+            // may optimize aligned writes depending on CIRCUITPY_FULL_BUILD of the CP build.
+            memcpy(dest_addr, values, len);
+            break;
+        case IO:
+            if ((size_t)dest_addr & 0x03 || len & 0x03) {
+                // Unaligned access or unaligned length not supported by RP2 for IO registers
+                mp_raise_RuntimeError(translate("Unable to access unaliged IO register"));
```
Comment (tannewt) [inline ports/raspberrypi/common-hal/memorymap/AddressRange.c:93]:
```suggestion
                mp_raise_RuntimeError(translate("Unable to access unaligned IO register"));
```

File: py/runtime.h [hunk 3]
```diff
@@ -45,6 +45,7 @@ typedef enum {
     MP_ARG_BOOL      = 0x001,
     MP_ARG_INT       = 0x002,
     MP_ARG_OBJ       = 0x003,
+    MP_ARG_UINT      = 0x004,
```
Comment (tannewt) [inline py/runtime.h:48]: Let's not modify `py/` at all. Anything we change is something we need to merge when we update. It looks like there is already [`mp_obj_int_get_uint_checked`](https://github.com/adafruit/circuitpython/blob/main/py/objint_mpz.c#L444-L458). (Added in https://github.com/micropython/micropython/commit/176ab99180a95eeac8794828f2f751a696571bb5) You can set the arg type to obj and then check it yourself in shared-bindings.

**Review by tannewt — CHANGES_REQUESTED** [review]
Thanks for the PR! Just a couple questions/suggestions.

File: ports/raspberrypi/common-hal/memorymap/AddressRange.c (hunk 1 above)
Comment (eightycc) [reply ports/raspberrypi/common-hal/memorymap/AddressRange.c:47]: Pushing back for these reasons: Overall, memorymap is a "beware, there be dragons here" sort of feature, so I prefer having a comprehensive view of the RP2 memory map and leaving its use to the discretion of the programmer. The other ports (espressif and nrf) that support memorymap do map RAM/SRAM/DRAM. Finally, it's useful to have a view into memory as a poor person's low-level debugger.

File: ports/raspberrypi/common-hal/memorymap/AddressRange.c (hunk 2 above)
Comment (eightycc) [reply ports/raspberrypi/common-hal/memorymap/AddressRange.c:93]: Ooops! I'll fix it.

File: py/runtime.h (hunk 3 above)
Comment (eightycc) [reply py/runtime.h:48]: Yes, there is already `mp_obj_int_get_uint_checked` and I'm using it, albeit indirectly. I'll kill `MP_ARG_UINT` and go with `MP_ARG_OBJ` as you suggest.

**Comment by eightycc** [comment]
I'll rebase this pull to the 8.2.x branch because that's how I'm already using it! Will GitHub automagically see the rebase or will I need to submit a new pull request?

File: locale/circuitpython.pot [hunk 4]
```diff
@@ -2127,6 +2136,14 @@ msgstr ""
 msgid "UUID value is not str, int or byte buffer"
 msgstr ""
 
+#: ports/raspberrypi/common-hal/memorymap/AddressRange.c
+msgid "Unable to access unaliged IO register"
```
Comment (tannewt) [inline locale/circuitpython.pot:2140]: Another `make translate` should remove this.

File: locale/circuitpython.pot [hunk 5]
```diff
@@ -4039,6 +4052,10 @@ msgstr ""
 msgid "source_bitmap must have value_count of 8"
 msgstr ""
 
+#: shared-bindings/memorymap/AddressRange.c
+msgid "start must be an int"
```
Comment (tannewt) [inline locale/circuitpython.pot:4056]: There should be helpers to verify this already. At minimum let's use the same error string. `%q` can be used in place of "start" in MP printf and passed the qstr.

**Review by tannewt — CHANGES_REQUESTED** [review]
Two more minor things. Thanks for the updates!

File: locale/circuitpython.pot (hunk 4 above)
Comment (eightycc) [reply locale/circuitpython.pot:2140]: Got it.

File: locale/circuitpython.pot (hunk 5 above)
Comment (eightycc) [reply locale/circuitpython.pot:4056]: There's not a helper for unsigned int types that I could find. I'd added one in an earlier commit as `mp_obj_get_uint`, but that involves a change to `py` which is undesirable. So, I've moved its code inside `memorymap_addressrange_make_new`. If you've got second thoughts about this approach, it's no trouble to move it back out to `py/obj.c`.

It turns out that the "start must be an int" message can be done away with entirely by attempting object conversion to an int by means of `mp_unary_op`, which I did with this commit.

**Review by tannewt — APPROVED** [review]
Looks good! Thank you!

**Pull request merged**